torch/nn/functional.py

```diff
@@ -3762,36 +3802,145 @@ def mse_loss(
+    Measures the element-wise mean squared error, with optional weighting.
+
+    Args:
+        input (Tensor): Predicted values.
+        target (Tensor): Ground truth values.
+        size_average (bool, optional): Deprecated (use reduction).
+        reduce (bool, optional): Deprecated (use reduction).
+        reduction (str, optional): Specifies the reduction to apply to the output:
+                                   'none' | 'mean' | 'sum'. 'mean': the mean of the output is taken.
+                                   'sum': the output will be summed. 'none': no reduction will be applied.
+                                   Default: 'mean'.
+        weights (Tensor, optional): Weights for each sample. Default: None.
+
+    Returns:
+        Tensor: Mean Squared Error loss (optionally weighted).
     """
     if has_torch_function_variadic(input, target):
         return handle_torch_function(
             mse_loss,
-            (input, target),
+            (input, target, size_average, reduce, reduction, weights),
             input,
             target,
             size_average=size_average,
             reduce=reduce,
             reduction=reduction,
+            weights=weights,
         )
+
     if not (target.size() == input.size()):
         warnings.warn(
             f"Using a target size ({target.size()}) that is different to the input size ({input.size()}). "
             "This will likely lead to incorrect results due to broadcasting. "
             "Please ensure they have the same size.",
             stacklevel=2,
         )
+
     if size_average is not None or reduce is not None:
         reduction = _Reduction.legacy_get_string(size_average, reduce)
 
-    expanded_input, expanded_target = torch.broadcast_tensors(input, target)
-    return torch._C._nn.mse_loss(
-        expanded_input, expanded_target, _Reduction.get_enum(reduction)
-    )
+    if weights is not None:
+        if weights.size() != input.size():
+            raise ValueError("Weights and input must have the same size.")
+
+        # Perform weighted MSE loss manually
+        expanded_input, expanded_target = torch.broadcast_tensors(input, target)
+        squared_errors = torch.pow(expanded_input - expanded_target, 2)
+        weighted_squared_errors = squared_errors * weights
+
+        if reduction == "none":
+            return weighted_squared_errors
+        elif reduction == "sum":
+            return torch.sum(weighted_squared_errors)
+        elif reduction == "mean":
+            return torch.sum(weighted_squared_errors) / torch.sum(weights)
+        else:
+            raise ValueError(
+                f"Invalid reduction mode: {reduction}. Expected one of 'none', 'mean', 'sum'."
+            )
+    else:
+        expanded_input, expanded_target = torch.broadcast_tensors(input, target)
+        return torch._C._nn.mse_loss(
+            expanded_input, expanded_target, _Reduction.get_enum(reduction)
+        )
+
+
+def mae_loss(
```

Comment:
Why are we adding mae_loss, isn't this synonymous to l1_loss? (from my previous comment, for wmae_loss we can add the weight argument to l1_loss), let me know if I missed something however :)